lib/PackageDb.js

```diff
@@ -0,0 +1,261 @@
+function traversePackageDb(
+  packageDb: PackageDb,
+  handler: (packageInfo: PackageInfo) => *,
+  packageName?: string = packageDb.rootPackageName
+) {
+  let packageInfo = packageDb.packagesByName[packageName];
+  let seen = new Set();
+  traversePackageDbImpl(
+    packageInfo,
+    seen,
+    packageDb,
+    handler
+  );
+}
+
+function traversePackageDbImpl(
+  packageInfo,
+  seen,
+  packageDb,
+  handler
+) {
+  let {packageJson, packageJsonFilePath} = packageInfo;
+  let dependencies = collectDependencies(packageDb, packageJson.name);
+  for (let depName of dependencies) {
+    if (seen.has(depName)) {
+      continue;
+    }
+    seen.add(depName);
+    let depPackageInfo = packageDb.packagesByName[depName];
+    traversePackageDbImpl(
+      depPackageInfo,
+      seen,
+      packageDb,
+      handler
+    );
+  }
+  handler(packageInfo)
+}
+
+function traversePackageTreeOnFileSystemSync(
+  packageJsonPathOnEjectingHost,
+  handler,
+  visitedRealPaths = {}
+) {
+  const packageJsonPathOnEjectingHostRealPath = fs.realpathSync(packageJsonPathOnEjectingHost);
+  const pkg = JSON.parse(fs.readFileSync(packageJsonPathOnEjectingHost, 'utf8'));
+  if (!pkg.name) {
+    throw ("no package name for package:" + packageJsonPathOnEjectingHost);
+  }
+  visitedRealPaths[pkg.name] = packageJsonPathOnEjectingHostRealPath;
+  /**
+   * How about the convention that `buildTimeOnlyDependencies` won't be
+   * traversed transitively to compute environments. The primary use case is
+   * that we generally only need a binary produced - or a dll.
+   */
+  KEYS.forEach((key) => {
+    Object.keys(pkg[key] || {}).forEach((dependencyName) => {
+      try {
+        const resolved = resolve.sync(
+          path.join(dependencyName, 'package.json'),
+          {basedir: path.dirname(packageJsonPathOnEjectingHost)}
+        );
+        if (!visitedRealPaths[dependencyName]) {
+          traversePackageTreeOnFileSystemSync(resolved, handler);
+        } else {
+          if (visitedRealPaths[dependencyName] !== fs.realpathSync(resolved)) {
+            // Find a way to aggregate warnings.
+            // console.warn(
+            //   "While computing environment for " + pkg.name + ", found that there are two separate packages named " +
+            //     dependencyName + " at two different real paths on disk. One is at " +
+            //     visitedRealPaths[dependencyName] + " and the other at " + fs.realpathSync(resolved)
+            // );
+          }
+        }
+      } catch (err) {
+        // We are forgiving on optional dependencies -- if we can't find them,
+        // just skip them
+        if (pkg["optionalDependencies"] && pkg["optionalDependencies"][dependencyName]) {
+          return;
+        }
+        throw err;
+      }
+    })
+  });
+  handler(packageJsonPathOnEjectingHost, pkg);
+}
+
+function normalizeName(name) {
+  return name
+    .toLowerCase()
+    .replace(/@/g, '')
+    .replace(/\//g, '_')
+    .replace(/\-/g, '_');
+}
```

Comment:
So normalised package name is lower cased as well as built-ins var names, examples:
* `$cur__root`
* `$packagea__lib`
* `$packagea__custom_var`
* ...